### PR TITLE
Passing a term to sparqlee to order

### PR DESCRIPTION
**Pull request by stephaniech97**

### CODE REVIEW

File: lib/Transformation.ts [hunk 1]
```diff
@@ -197,3 +197,24 @@ export function transformAggregate(expr: Alg.AggregateExpression) {
 export function transformExistence(expr: Alg.ExistenceExpression) {
   return new E.Existence(expr);
 }
+
+// Ordering 2 given strings
+export function orderTypes(litA: RDF.NamedNode | RDF.BlankNode | RDF.Literal | RDF.Variable | RDF.DefaultGraph | undefined,
```
Comment (rubensworks) [inline lib/Transformation.ts:202]: `RDF.NamedNode | RDF.BlankNode | RDF.Literal | RDF.Variable | RDF.DefaultGraph` can be simplified to `RDF.Term`.

Also, I would move these two new functions into a new file (`Ordering.ts`), as this is not really a transformation.

File: lib/Transformation.ts [hunk 2]
```diff
@@ -197,3 +197,24 @@ export function transformAggregate(expr: Alg.AggregateExpression) {
...
+
+// Ordering 2 given strings
+export function orderTypes(litA: RDF.NamedNode | RDF.BlankNode | RDF.Literal | RDF.Variable | RDF.DefaultGraph | undefined, 
+                        litB: RDF.NamedNode | RDF.BlankNode | RDF.Literal | RDF.Variable | RDF.DefaultGraph | undefined, 
+                        isAscending:boolean){
+  if (litA && litA.termType === "Literal" && litB && litB.termType === "Literal" ){
+    const a = transformLiteral(litA);
+    const b = transformLiteral(litB); 
+    return order(a, b, isAscending);
+  }else{
```
Comment (rubensworks) [inline lib/Transformation.ts:209]: Missing spaces around `else`.

File: lib/Transformation.ts [hunk 3]
```diff
@@ -197,3 +197,24 @@ export function transformAggregate(expr: Alg.AggregateExpression) {
 export function transformExistence(expr: Alg.ExistenceExpression) {
   return new E.Existence(expr);
 }
+
+// Ordering 2 given strings
+export function orderTypes(litA: RDF.NamedNode | RDF.BlankNode | RDF.Literal | RDF.Variable | RDF.DefaultGraph | undefined, 
+                        litB: RDF.NamedNode | RDF.BlankNode | RDF.Literal | RDF.Variable | RDF.DefaultGraph | undefined, 
+                        isAscending:boolean){
+  if (litA && litA.termType === "Literal" && litB && litB.termType === "Literal" ){
```
Comment (rubensworks) [inline lib/Transformation.ts:205]: We typically write our strings with single quotes (our linter is probably not configured strictly enough...).

File: lib/util/Consts.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 // tslint:disable:variable-name
 import * as RDFDM from '@rdfjs/data-model';
-import { Map, Set } from 'immutable';
+import { Map, Set, is } from 'immutable';
```
Comment (rubensworks) [inline lib/util/Consts.ts:3]: Unneeded import?

File: lib/Transformation.ts (hunk 1 above)
Comment (rubensworks) [reply lib/Transformation.ts:202]: I'm also wondering now if literals can really be undefined. What happens if you remove these undefined-checks?

Comment (stephaniech97) [reply lib/Transformation.ts:202]: const transform = async (bindings: Bindings, next: any) => {
        try {
          const result = await evaluator.evaluate(bindings);
          transformedStream._push({ bindings, result });
          
        } catch (err) {
          if (!isExpressionError(err)) {
            bindingsStream.emit('error', err);
          }
          transformedStream._push({ bindings, result: undefined });
        }
        next();
      };
In the ActorQueryOperationOrderBySparqlee there is a possibility of having an error. If for example I would pass a non-literal into the evaluate it will throw an error and then still do the ordering but with undefined.

File: lib/util/Consts.ts (hunk 4 above)
Comment (stephaniech97) [reply lib/util/Consts.ts:3]: The import was never used

File: lib/Transformation.ts [hunk 5]
```diff
@@ -110,7 +110,6 @@ export function transformLiteral(lit: RDF.Literal): E.Literal<any> {
 
     case DT.XSD_INTEGER:
     case DT.XSD_DECIMAL:
-
```
Comment (rubensworks) [inline lib/Transformation.ts:113]: Could you undo the changes in this file? That will make the diff a bit cleaner :-)

File: test/util.test/Parsing.test.ts [hunk 6]
```diff
@@ -54,4 +54,4 @@ describe('util/parsing', () => {
       });
     });
   });
-});
+});
```
Comment (rubensworks) [inline test/util.test/Parsing.test.ts:57]: Changes here can also be undone.

Also, it looks like your editor removes empty lines at the end of files, while these are typically good to have. Maybe a config option somewhere in your editor?

File: lib/util/Ordering.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        const b = transformLiteral(litB); 
+        return order(a, b, isAscending);
+    } else {
+        return 0;
+    }
+}
+
+// Effective ordering
+export function order(orderA: T.Literal<any>, orderB: T.Literal<any>, isAscending:boolean){
+    if (!orderA || !orderB || orderA.typedValue === orderB.typedValue) {
```
Comment (rubensworks) [inline lib/util/Ordering.ts:18]: As far as I can see, the output of `transformLiteral` is always defined. So I think these falsy checks can be removed, right?

File: lib/util/Ordering.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import * as RDF from 'rdf-js';
+import * as T from '../expressions/Term';
+import { transformLiteral } from '../Transformation';
+
+// Ordering 2 given strings
```
Comment (rubensworks) [inline lib/util/Ordering.ts:5]:
```suggestion
// Determine the relative numerical order of the two given terms.
```

**Comment by rubensworks** [comment]
Great work, thanks!
